File: 2.0/LinFu.DesignByContract2.Examples/ContractAssertions/Assertions/NonNegativeAttribute.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Reflection;
using System.Text;
using LinFu.DesignByContract2.Attributes;
using LinFu.DesignByContract2.Core;
using LinFu.DynamicProxy;

namespace CSContractAssertions
{
    [AttributeUsage(AttributeTargets.Parameter | AttributeTargets.ReturnValue)]
    public class NonNegativeAttribute : Attribute, IParameterPrecondition, IPostcondition 
    {
        #region IParameterPrecondition Members

        public bool Check(InvocationInfo info, ParameterInfo parameter, object argValue)
        {
            int value = Convert.ToInt32(argValue);
            return value >= 0;
        }

        public void ShowErrorMessage(TextWriter stdOut, InvocationInfo info, 
            ParameterInfo parameter, object argValue)
        {
            stdOut.WriteLine("The parameter '{0}' must be non-negative", parameter.Name);
        }

        #endregion

        #region IContractCheck Members

        public bool AppliesTo(object target, InvocationInfo info)
        {
            if (target == null)
                return false;

            return true;           
        }

        public void Catch(Exception ex)
        {
            // Ignore the error
        }

        #endregion

        #region IPostcondition Members

        public void BeforeMethodCall(object target, InvocationInfo info)
        {
        }

        public bool Check(object target, InvocationInfo info, object returnValue)
        {
            int value = Convert.ToInt32(returnValue);
            return value >= 0;
        }

        public void ShowError(TextWriter output, object target, InvocationInfo info, object returnValue)
        {
            output.WriteLine("The return value must be non-negative");
        }

        #endregion
    }
}

File: 2.0/LinFu.DesignByContract2.Examples/VBContractAssertions/Assertions/NonNegativeAttribute.vb
Imports LinFu.DesignByContract2.Attributes
Imports LinFu.DesignByContract2.Core

<AttributeUsage(AttributeTargets.Parameter Or AttributeTargets.ReturnValue)> _
Public Class NonNegativeAttribute
    Inherits Attribute
    Implements IParameterPrecondition
    Implements IPostcondition

    Public Function Check(ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal parameter As System.Reflection.ParameterInfo, ByVal argValue As Object) As Boolean Implements LinFu.DesignByContract2.Attributes.IParameterPrecondition.Check
        Dim value As Integer = Convert.ToInt32(argValue)
        Return value >= 0
    End Function

    Public Sub ShowErrorMessage(ByVal stdOut As System.IO.TextWriter, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal parameter As System.Reflection.ParameterInfo, ByVal argValue As Object) Implements LinFu.DesignByContract2.Attributes.IParameterPrecondition.ShowErrorMessage
        stdOut.WriteLine("The parameter '{0}' must be non-negative", parameter.Name)
    End Sub

    Public Function AppliesTo(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo) As Boolean Implements LinFu.DesignByContract2.Core.IContractCheck.AppliesTo
        If target Is Nothing Then Return False
        Return True
    End Function

    Public Sub [Catch](ByVal ex As System.Exception) Implements LinFu.DesignByContract2.Core.IContractCheck.Catch
        ' Do nothing
    End Sub

    Public Sub BeforeMethodCall(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo) Implements LinFu.DesignByContract2.Core.IPostcondition.BeforeMethodCall

    End Sub

    Public Function Check1(ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal returnValue As Object) As Boolean Implements LinFu.DesignByContract2.Core.IPostcondition.Check
        Dim value As Integer = Convert.ToInt32(returnValue)
        Return value >= 0
    End Function

    Public Sub ShowError(ByVal output As System.IO.TextWriter, ByVal target As Object, ByVal info As LinFu.DynamicProxy.InvocationInfo, ByVal returnValue As Object) Implements LinFu.DesignByContract2.Core.IPostcondition.ShowError
        output.WriteLine("The return value must be non-negative")
    End Sub
End Class
